### PR TITLE
feat: Hooks for in-process user chaincode

### LINKED ISSUE #158: Hooks for in-process user chaincode
Add hooks for enabling in-process user chaincode.

### PR DESCRIPTION
**Pull request by bstasyszyn**
Add hooks that support in-process user chaincode extensions.

closes #158

Signed-off-by: Bob Stasyszyn <Bob.Stasyszyn@securekey.com>

### CODE REVIEW

File: extensions/go.sum [hunk 1]
```diff
@@ -30,6 +30,7 @@ github.com/beorn7/perks v0.0.0-20180321164747-3a771d992973/go.mod h1:Dwedo/Wpr24
 github.com/beorn7/perks v1.0.0/go.mod h1:KWe93zE9D1o94FZ5RNwFwVgaQK1VOXiVxmqh+CedLV8=
 github.com/beorn7/perks v1.0.1/go.mod h1:G2ZrVWU2WbWT9wwq4/hrbKbnv/1ERSJQ0ibhJ6rlkpw=
 github.com/bgentry/speakeasy v0.1.0/go.mod h1:+zsyZBPWlz7T6j88CTgSN5bM796AkVf0kBD4zp0CCIs=
+github.com/bstasyszyn/fabric-protos-go-ext v0.0.0-20191212155638-f5cf9faefce1/go.mod h1:xVYTjK4DtZRBxZ2D9aE4y6AbLaPwue2o/criQyQbVD0=
```
Comment (troyronda) [inline extensions/go.sum]: This looks like it need a tidy.

Comment (bstasyszyn) [reply extensions/go.sum]: Done

File: core/ledger/kvledger/txmgmt/privacyenabledstate/common_storage_db.go [hunk 2]
```diff
@@ -335,6 +336,52 @@ func (s *CommonStorageDB) HandleChaincodeDeploy(chaincodeDefinition *cceventmgmt
...
+	if !ok {
+		logger.Debugf("No index definitions for DB of type [%s] for chaincode [%s]", indexCapable.GetDBType(), chaincodeDefinition.Name)
+		return nil
+	}
+
+	collectionConfigMap, err := extractCollectionNames(chaincodeDefinition)
+	if err != nil {
+		logger.Errorf("Error while retrieving collection config for chaincode=[%s]: %s",
+			chaincodeDefinition.Name, err)
+		return nil
```
Comment (sandrask) [inline core/ledger/kvledger/txmgmt/privacyenabledstate/common_storage_db.go]: is this supposed to be err?

File: core/ledger/kvledger/txmgmt/privacyenabledstate/common_storage_db.go [hunk 3]
```diff
@@ -335,6 +336,52 @@ func (s *CommonStorageDB) HandleChaincodeDeploy(chaincodeDefinition *cceventmgmt
...
+	}
+
+	logger.Debugf("Adding %d index(es) for chaincode [%s]", len(dbArtifacts.Indexes), chaincodeDefinition.Name)
+	if err := indexCapable.ProcessIndexes(chaincodeDefinition.Name, dbArtifacts.Indexes); err != nil {
+		logger.Errorf("Error processing index for chaincode [%s]: %s", chaincodeDefinition.Name, err)
```
Comment (sandrask) [inline core/ledger/kvledger/txmgmt/privacyenabledstate/common_storage_db.go]: no return here, is this OK?

File: core/ledger/kvledger/txmgmt/privacyenabledstate/common_storage_db.go (hunk 2 above)
Comment (bstasyszyn) [reply core/ledger/kvledger/txmgmt/privacyenabledstate/common_storage_db.go]: I copied the logic from HandleChaincodeDeploy, which is also logging an error but returning nil. Not sure why they do this but I didn't want to change the behaviour.

Comment (bstasyszyn) [reply core/ledger/kvledger/txmgmt/privacyenabledstate/common_storage_db.go]: On second thought, I think there should be an error returned. I'll change it.

File: core/ledger/kvledger/txmgmt/privacyenabledstate/common_storage_db.go (hunk 3 above)
Comment (bstasyszyn) [reply core/ledger/kvledger/txmgmt/privacyenabledstate/common_storage_db.go]: Same here - I copied the existing logic but I don't think it's right. I'll change it to return an error.

**Pull request merged**